theano/sandbox/cuda/dnn.py

```diff
@@ -2186,6 +2186,460 @@ def method(self):
+        result = ''
+        for id in self.tensor_4d_descs:
+            result += self._define_tensor4d_desc(name, id)
+        return result
+
+    def c_init_code_struct(self, node, name, sub):
+        result = """
+cudnnStatus_t err%(name)s;
+""" % dict(name=name)
+
+        for id in self.tensor_4d_descs:
+            result += self._init_tensor4d_desc(name, id, sub['fail'])
+        return result
+
+    def c_cleanup_code_struct(self, node, name):
+        result = ''
+        for id in self.tensor_4d_descs:
+            result += self._clean_tensor4d_desc(name, id)
+        return result
+
+    def c_code(self, node, name, inputs, outputs, sub):
+        if self.tensor_format == 'b01c':
+            tensor_format = 1
+        else:
+            tensor_format = 0
+
+        if self.mode == "spatial":
+            mode = "CUDNN_BATCHNORM_SPATIAL"
+        else:
+            mode = "CUDNN_BATCHNORM_PER_ACTIVATION"
+
+        # Setup configuration variables.
+        result = """
+cudnnStatus_t err%(name)s;
+cudnnTensorFormat_t format%(name)s = CUDNN_TENSOR_NCHW;
+if (%(tensor_format)d == 1)
+  format%(name)s = CUDNN_TENSOR_NHWC;
+
+cudnnBatchNormMode_t mode%(name)s = %(mode)s;
+
+double exponentialAverageFactor%(name)s = %(exp_avg_factor)f;
+
+double epsilon%(name)s = %(epsilon)e;
+""" % dict(name=name, tensor_format=tensor_format, mode=mode, exp_avg_factor=self.exp_avg_factor,
+           epsilon=self.epsilon)
+
+        # Validate the input and build the input variables.
+        for input_idx, bn_input in enumerate(inputs):
+            if self.direction == 'backward':
+                if input_idx == 0:
+                    input_name = 'bn_input'
+                elif input_idx == 1:
+                    input_name = 'bn_gout'
+                else:
+                    input_name = 'scale_bias_mean'
+            else:
+                if input_idx == 0:
+                    input_name = 'bn_input'
+                else:
+                    input_name = 'scale_bias_mean'
+            result += c_set_tensor4d(bn_input, input_name + "_" + name,
+                                     "err" + name, sub['fail'])
+
+        subs = dict(fail=sub['fail'], name=name)
+
+        for idx, bn_input in enumerate(inputs):
+            subs['ins%d' % idx] = bn_input
+
+        # Build and prepare the output variable.
+        sub_result = ""
+        for idx, bn_output in enumerate(outputs):
+            subs['outs%d' % idx] = bn_output
+            sub_result += """
+    if (CudaNdarray_prep_output(&%(outs""" + str(idx) + """)s, 4, CudaNdarray_HOST_DIMS(%(ins""" + str(idx) + """)s)) != 0)
+    {
+      %(fail)s
+    }
+    """
+        result += sub_result % subs
+
+        for output_idx, output in enumerate(outputs):
+            if self.direction == 'backward':
+                if output_idx == 0:
+                    output_name = 'bn_ginput'
+                else:
+                    output_name = 'scale_bias_mean'
```

Comment:
There is more then 2 outputs for the grad, ans all index 1-N seem to be sharing the same dnn descriptor here.